### PR TITLE
Add WASAPI loopback implementation

### PR DESCRIPTION
**Pull request by WeekendWarrior1**
Re implements this patch: https://github.com/audacity/audacity/commit/be7467c0e32c397d6223f2dc844a585d079b9875#diff-6ea9c8ba5ecef609fc4cfa9fbbb1bc11ad66acb997565df06ad6ce1562a10144 to add loopback for WASAPI devices.

I'm very new to C so I welcome any feedback or improvements to this PR.

Closes this:
PortAudio/portaudio#668

And would really help python-sounddevice users:
spatialaudio/portaudio-binaries#6
spatialaudio/python-sounddevice#287
https://github.com/tlecomte/friture/issues/64
https://github.com/spatialaudio/python-sounddevice/issues/281
spatialaudio/python-sounddevice#14 

Also see this project which maintains a fork of portaudio with loopback support specifically for PyAudio: https://github.com/intxcc/pyaudio_portaudio

### CODE REVIEW

**Comment by RossBencina** [comment]
@dmitrykos could you review this please

**Comment by WeekendWarrior1** [comment]
This was marked as a draft because shauneccles discovered some issues when using other devices after testing this patch, will bug hunt tonight.

**Comment by dmitrykos** [comment]
@RossBencina yes, I will do so once have a bit of free time, I am tracking this PR already since it was proposed. I also will wait for announced @WeekendWarrior1's improvements.

File: src/hostapi/wasapi/pa_win_wasapi.c [hunk 1]
```diff
@@ -478,6 +478,9 @@ typedef struct PaWasapiDeviceInfo
 
     // Form-factor
     EndpointFormFactor formFactor;
+
+	// Loopback indicator
+	int loopBack;
```
Comment (dmitrykos) [inline src/hostapi/wasapi/pa_win_wasapi.c:483]: Please use BOOL type.

File: src/hostapi/wasapi/pa_win_wasapi.c [hunk 2]
```diff
@@ -2266,6 +2323,34 @@ static PaError CreateDeviceList(PaWasapiHostApiRepresentation *paWasapi, PaHostA
...
+    UINT32 initialDevices = paWasapi->deviceCount;
+    for (i = 0; i < initialDevices; ++i) {
+        if (paWasapi->devInfo[i].flow == eRender) {
+            // need to add a loopback for this render device
+            if ((deviceInfoArray = AddLoopbacksToDeviceList(paWasapi, deviceInfoArray, i)) == NULL)
+            {
+                result = paInsufficientMemory;
+                goto error;
+            }
+            paWasapi->devInfo[paWasapi->deviceCount].loopBack = 1;
```
Comment (dmitrykos) [inline src/hostapi/wasapi/pa_win_wasapi.c:2336]: Correct to `= TRUE` instead of `= 1`.

File: src/hostapi/wasapi/pa_win_wasapi.c [hunk 3]
```diff
@@ -3858,7 +3965,8 @@ static PaError OpenStream( struct PaUtilHostApiRepresentation *hostApi,
             (stream->in.shareMode == AUDCLNT_SHAREMODE_SHARED) &&
             ((inputStreamInfo != NULL) && (inputStreamInfo->flags & paWinWasapiAutoConvert)))
             stream->in.streamFlags |= (AUDCLNT_STREAMFLAGS_AUTOCONVERTPCM | AUDCLNT_STREAMFLAGS_SRC_DEFAULT_QUALITY);
-
+        if (info->flow == eRender)
+            stream->in.streamFlags |= AUDCLNT_STREAMFLAGS_LOOPBACK;
```
Comment (dmitrykos) [inline src/hostapi/wasapi/pa_win_wasapi.c:3969]: Please separate this check with spaces and add proper comment which explains this flag. `if (info->flow == eRender)` must be replaced with `if (info->loopBack)`.

**Comment by RossBencina** [comment]
Awaiting response to Dmitry's change requests.

**Comment by dmitrykos** [comment]
@not-matt thank you for the changes.

The implementation can be improved further by calculating the number of available `eRender` devices in order to allocate memory once here:
https://github.com/PortAudio/portaudio/blob/2a68aeb5845d248daaec556af4de6c8d77752f26/src/hostapi/wasapi/pa_win_wasapi.c#L2294 Otherwise memory is wasted (not leaked but it is inefficient anyway) when you reassign it here: https://github.com/PortAudio/portaudio/blob/2a68aeb5845d248daaec556af4de6c8d77752f26/src/hostapi/wasapi/pa_win_wasapi.c#L2343

If memory is allocated once by `AllocateDeviceListMemory` then `AddLoopbacksToDeviceList` could be minimized to only fill the loopback-specific info inside the https://github.com/PortAudio/portaudio/blob/2a68aeb5845d248daaec556af4de6c8d77752f26/src/hostapi/wasapi/pa_win_wasapi.c#L2301 loop.

It would make implementation much simpler.

Here is helper function which will return the number of eRender devices in the device list:

```c
// ------------------------------------------------------------------------------------------
static UINT32 GetRenderDeviceCount(IMMDeviceCollection *pEndPoints)
{
    HRESULT hr;
    UINT32 deviceCount;
    IMMDevice *device;
    IMMEndpoint *endpoint;
    EDataFlow flow;
    UINT32 ret = 0;

    hr = IMMDeviceCollection_GetCount(pEndPoints, &deviceCount);
    IF_FAILED_JUMP(hr, error);

    for (UINT32 i = 0; i < deviceCount; ++i)
    {
        hr = IMMDeviceCollection_Item((IMMDeviceCollection *)pEndPoints, i, &device);
        IF_FAILED_JUMP(hr, error);

        if (SUCCEEDED(hr = IMMDevice_QueryInterface(device, &pa_IID_IMMEndpoint, (void **)&endpoint)))
        {
            if (SUCCEEDED(hr = IMMEndpoint_GetDataFlow(endpoint, &flow)))
                ret += (flow == eRender);

            SAFE_RELEASE(endpoint);
        }

        SAFE_RELEASE(device);
    }

    return ret;

error:

    return 0;
}
```

**Comment by not-matt** [comment]
Hi @dmitrykos, thank you for the feedback. I've worked with @WeekendWarrior1 in the past but I've not been able to reach him to wrap up this PR. 

I can make an attempt to implement your feedback, but I have no experience writing C. If you have the time, would you consider writing these changes yourself? 

I'm in touch with [@shauneccles](https://github.com/shauneccles) and can get any changes you would make added to this PR very quickly.

**Comment by dmitrykos** [comment]
@not-matt ok, I will try to implement in the way I advised and let's see how it works then.

**Comment by not-matt** [comment]
Nice, cheers. Do let me know if there's anything I can help you with, especially testing.

**Comment by nidefawl** [comment]
I have been waiting for this feature to land upstream. I maintained my own version for a couple of years. 
I will help testing once @dmitrykos is done with the implementation

**Comment by dmitrykos** [comment]
I made necessary cleanup changes, would you please test how it works on your side. I tested Win32 and UWP modes and it works quite well, no problems noted. 

Loopback mode does not support Exclusive mode though that is quite natural because it is a proxy for a playback device.

If your test results confirm that all runs fine I think we could finally merge it.

**Comment by not-matt** [comment]
Good stuff! I'm away this weekend, I'll take a look once I'm back.

**Comment by nidefawl** [comment]
I built and linked against this PR and the loopback input works nicely!

I haven't tested the `PaWasapi_IsLoopback` function yet

**Comment by nidefawl** [comment]
PaWasapi_IsLoopback works as intended

**Comment by dmitrykos** [comment]
@nidefawl great, thank you for testing!

**Comment by shauneccles** [comment]
Can confirm it behaves as expected over here on Win 10/Win 11.

**Comment by dmitrykos** [comment]
@shauneccles thank you for testing. I added comments into the source code regarding `[Loopback]` loopback device name identificator usage. If the form of this identificator is unexpected by some of your projects, please comment. I assumed that Windows audio device names often have device description in (), so to differentiate identificator more I used `[Loopback]`. It can be adjusted of course if such change created some problems.

**Comment by shauneccles** [comment]
No issues with using square brackets that I can see/think of, it makes sense to keep them different to give people another signal that it is a loopback device.

**Comment by not-matt** [comment]
+1, this works well for me on Windows. Haven't ran into any issues.

**Comment by dmitrykos** [comment]
Thank you for proposing the PR and tests, I think it can go into the main repo!

**Comment by bear101** [comment]
After adding loopback devices to WASAPI it looks like PaHostApiInfo::defaultInputDevice and PaHostApiInfo::defaultOutputDevice are broken. I haven't quite figured out why the indexing is not matching.
```
#include <portaudio.h>
#include <iostream>

int main(int argc, char* args[])
{
    PaError err = Pa_Initialize();

    int n_devices = Pa_GetDeviceCount();

    for (int i=0;i<n_devices;++i)
    {
        const PaDeviceInfo* devinfo = Pa_GetDeviceInfo(i);
        std::cout << "#" << i << " " << devinfo->name << std::endl;
    }

    PaHostApiIndex hostapiIndex = Pa_HostApiTypeIdToHostApiIndex(paWASAPI);
    const PaHostApiInfo* hostapi = Pa_GetHostApiInfo(hostapiIndex);

    std::cout << "Default WASPI input #" << hostapi->defaultInputDevice << std::endl;
    std::cout << "Default WASPI output #" << hostapi->defaultOutputDevice << std::endl;
    return 0;
}
```
Output:
```
#0 Microsoft Sound Mapper - Input
#1 Microphone (High Definition Aud
#2 Microphone (HD Pro Webcam C920)
#3 Microsoft Sound Mapper - Output
#4 Headphones (High Definition Aud
#5 VE248 (Intel(R) Display Audio)
#6 Primary Sound Capture Driver
#7 Microphone (High Definition Audio Device)
#8 Microphone (HD Pro Webcam C920)
#9 Primary Sound Driver
#10 Headphones (High Definition Audio Device)
#11 VE248 (Intel(R) Display Audio)
#12 VE248 (Intel(R) Display Audio)
#13 Headphones (High Definition Audio Device)
#14 Microphone (High Definition Audio Device)
#15 Microphone (HD Pro Webcam C920)
#16 VE248 (Intel(R) Display Audio) [Loopback]
#17 Headphones (High Definition Audio Device) [Loopback]
Default WASPI input #16
Default WASPI output #14
```
Something must be going wrong with the indexing either [here](https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wasapi/pa_win_wasapi.c#L1982) and [here](https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wasapi/pa_win_wasapi.c#L2383).

**Comment by dmitrykos** [comment]
@bear101 thank you for reporting, I will have a look shortly.

**Comment by dmitrykos** [comment]
@bear101 would you please check #688 PR regarding the problem you noticed.

**Comment by dechamps** [comment]
@GiuseppeUFSM You don't need any special code to use this feature. Just use PortAudio normally and the "virtual" loopback devices will just naturally show up in the usual list of input devices. In fact it should even work in *existing* applications if you replace their PortAudio DLL.

**Pull request merged**